CrossPare/src/main/java/de/ugoe/cs/cpdp/dataselection/OutOfSampleBootstrap.java

```diff
@@ -0,0 +1,152 @@
+package de.ugoe.cs.cpdp.dataselection;
+
+import de.ugoe.cs.cpdp.util.WekaUtils;
+import de.ugoe.cs.cpdp.versions.SoftwareVersion;
+import org.apache.commons.collections4.list.SetUniqueList;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import weka.core.Instance;
+import weka.core.Instances;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+/**
+ * Applies Out-Of-Sample Bootstrap to the data.
+ * <p>
+ * With this data selection technique a bootstrap sample from the data is drawn and used as trainings data.
+ * The leftover instances, which are not in the bootstrap sample, are used as trainings data.
+ * Statistically these are 36.8 % of the data.
+ * If there are not at least a minimal number of samples in the drawn trainings data the procedure is repeated.
+ *
+ * @author Steffen Tunkel
+ */
+public class OutOfSampleBootstrap implements IPointWiseDataselectionStrategy {
+
+    /**
+     * Reference to the logger
+     */
+    private static final Logger LOGGER = LogManager.getLogger("main");
+
+    /**
+     * minimum number of defects in a sample to keep it
+     */
+    private int minDefects = 2;
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see de.ugoe.cs.cpdp.IParameterizable#setParameter(java.lang.String)
+     */
+    @Override
+    public void setParameter(String parameters) {
+        if (parameters != null && !parameters.isEmpty()) {
+            this.minDefects = Integer.parseInt(parameters);
+        }
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see de.ugoe.cs.cpdp.dataselection.IPointWiseDataselectionStrategy#apply(de.ugoe.cs.cpdp.versions.SoftwareVersion,
+     * de.ugoe.cs.cpdp.versions.SoftwareVersion)
+     */
+    @Override
+    public SoftwareVersion apply(SoftwareVersion testversion, SoftwareVersion trainversion) {
+        WekaUtils.makeClassBinary(testversion.getInstances());
+        WekaUtils.makeClassBinary(trainversion.getInstances());
```

Comment:
avoid side effects. rather let this crash with an exception if the class is not yet binary.